### ISSUE TITLE
Does not open last used database automatically

### ISSUE THREAD

**Issue by mase76** [keepassxreboot/keepassxc#4911]
Hi!
I often have the issue, that when using the global autotype shortcut, nothing happens, because KeepassXC does not open the last used database. I have to open the main window and select the last (and only) database from the recents list. Then it works for a while. Maybe there should be an option to set the default database.

**Comment by droidmonkey** [comment]
I'm confused by the problem. Do you even have the database open in keepassxc? We don't open a database file, you need to do that first.

**Comment by mase76** [comment]
Sorry, I did not mean opened and decrypted, but selected with its keyfile, so that you only have to type the passphrase.

**Comment by droidmonkey** [comment]
Do you have "remember last databases and key files" setting set?

![image](https://user-images.githubusercontent.com/2809491/85958395-3f1d3100-b963-11ea-918c-6bed83c1cb7d.png)

**Comment by mase76** [comment]
Yes, I have. But it seems, that it sometimes does not know, what the last db was. I have this issue on multiple machines.

**Comment by droidmonkey** [comment]
I am going to need screenshots or a demo video to understand what your problem is, sorry.

**Comment by mase76** [comment]
Ok, I attached a screencast. I am trying to log into Github via autotype, but nothing happens, because the last used database was not remembered after a reboot. I have to open the KeepassXC window and select my database from the list of the last opened (I have only one database).This does not happen every time, but very often. The expected behabiour is, that my last used database it opened every time, and I am only prompted to unlock it.
[keepassxc.tar.gz](https://github.com/keepassxreboot/keepassxc/files/4847574/keepassxc.tar.gz)

**Comment by droidmonkey** [comment]
When you close keepassxc do not close the database (don't press Ctrl+W). That will require you to open it again on launching keepassxc.

**Comment by mase76** [comment]
I do not close the db. The issue is mostly after reboots or resumes.

**Comment by droidmonkey** [comment]
Is your db hosted on a network shares that might not be connected or available?

**Comment by mase76** [comment]
No, the db file is always present in the locsl filesystem.
I suggest, when using the autotype shortcut, and only one recent db file is in the list, always to use this one. Or opening the application to let me choose the db, instead of doing nothing.
My wife also has this issue. Maybe it is related to resuming from hibernation. I did not have this somewhere in the past. But I cannot say, when it appeared for the first time, or with the move from KeepassX to KeepassXC.
But when you or anybody else never had this issue, it seems to be my fault.

**Comment by rplCloud** [comment]
I was facing the same problem.

In the settings GUI the option "Load previously used database" was set but in the `~/Library/Application Support/keepassxc/keepassxc.ini` file the option `OpenPreviousDatabasesOnStartup` was set to `false`.

I changed the value to `true` and chose a file from "Recent databases". It works fine ever since...

**Comment by SaraSmiseth** [comment]
I have the same problem. After a reboot keepass forgets what my last database was. As long as I don't reboot it works fine. I have the GUI Settings enabled and OpenPreviousDatabasesOnStartup is also set to true.

**Comment by droidmonkey** [comment]
This is a common theme and unfortunate finding

**Comment by ycaille** [comment]
I have the same issue. 

Despite "_Remember previously used databases_", "_Load previously open databases on startup_" and "_Remember database key files and security dongles_" being checked in "**Application Settings > Basic Settings**", whenever I quit the application to reboot the system, I have to provide explicitly the name of both DB and key when I open KeePassXC. Also, the settings are correctly saved in the **keepassxc.ini** file

I've never had the problem before v2.6.0, and I've been using KeePassXC for years.

It does not happen when I just quit the application (Cmd+Q).

Both files (DB & key) are located on a USB drive, but they are both available when I start KeePassXC.

OS: macOS Mojave 10.14.6
<img width="610" alt="Screen Shot 2020-07-17 at 1 48 15 PM" src="https://user-images.githubusercontent.com/1269410/87788279-628a1c00-c83d-11ea-9210-658a2545d59d.png">
<img width="281" alt="Screen Shot 2020-07-17 at 2 53 45 PM" src="https://user-images.githubusercontent.com/1269410/87788286-67e76680-c83d-11ea-9cf8-7b1eb09de03c.png">

**Comment by droidmonkey** [comment]
Sounds like a completely different issue. Please open a new ticket for that

**Comment by z3ntu** [comment]
Same for me, `~/.cache` is on tmpfs and therefore wiped after a reboot. This is new behavior since v2.8.0, it was always working fine before that.

**Comment by droidmonkey** [comment]
Technically we are not doing anything wrong, mounting `.cache` to a tempfs is not a best practice that we are aware of. Cache is meant to survive reboots, but not be synced with other machines. We are using standard Qt directories for "local config" storage, so your actions to mount to a tempfs may impact other applications function.

**Comment by SaraSmiseth** [comment]
> Technically we are not doing anything wrong, mounting `.cache` to a tempfs is not a best practice that we are aware of. Cache is meant to survive reboots, but not be synced with other machines. We are using standard Qt directories for "local config" storage, so your actions to mount to a tempfs may impact other applications function.

Yes you are right. This is not a problem of keepassxc. I have other applications that save configs in `.cache`. I save them somewhere else and just symlink them to `.cache` after each start.

For example I use this for keepassxc:

``` sh
mkdir -p ~/.cache/keepassxc && ln -s ~/keepassxc.ini ~/.cache/keepassxc/keepassxc.ini
```

**Comment by z3ntu** [comment]
> There is a single base directory relative to which user-specific non-essential (cached) data should be written. This directory is defined by the environment variable $XDG_CACHE_HOME.

> $XDG_CACHE_HOME defines the base directory relative to which user specific non-essential data files should be stored.

https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

Imo the "last opened" list is not non-essential data and should go into XDG_DATA_HOME instead.

**Comment by droidmonkey** [comment]
Yes the change is deliberate and the directory choice is correct

**Comment by z3ntu** [comment]
@droidmonkey I don't really understand why the change was done though. Were there problems with ~/.config/keepassxc ?

**Comment by droidmonkey** [comment]
We store settings in both places. Ephemeral, system specific settings are stored in the "local" storage, which on Linux corresponds to the .cache directory. All other settings remain in .config

**Comment by z3ntu** [comment]
Okay. I do understand that saving a "recently used" list in ~/.config is "wrong". I don't understand why ~/.cache instead of ~/.local/share was chosen though as the latter would make much more sense (to me at least - it's definitely not "non-essential data" in my eyes).
So replacing `QStandardPaths::GenericCacheLocation` with `QStandardPaths::AppDataLocation` at https://github.com/keepassxreboot/keepassxc/blob/5c2a1a4284c1bccfcd99a861336559105ed3db7d/src/core/Config.cpp#L449
should make everyone happy?

**Comment by droidmonkey** [comment]
Ping @phoerious

**Comment by gitvalds** [comment]
On Fedora 35 /tmp/ is a tmpfs.

The following procedure works for me on Fedora 35 with KDE, for replacing:
```${HOME}/.cache```
with:
```/tmp/kde_cache/```
but preserving the settings of keepassxc.

You can copy paste after replacing all the occurrencies of:
```<user>```
with the name of your user.
Thank you so much to everybody who contributes to this amazing software :+1:
_____________________________________________________

As the user ```<user>```, run the following commands:
```
mv /home/<user>/.cache /home/<user>/.cache.renamed
mkdir /tmp/kde_cache/
ln -s /tmp/kde_cache/ /home/<user>/.cache
mkdir /home/<user>/.local/share/keepassxc
cp /home/<user>/.cache.renamed/keepassxc/keepassxc.ini /home/<user>/.local/share/keepassxc/keepassxc.ini
```
_____________________________________________________

As the user root, run the following commands:
```
vim /etc/profile.d/my.sh
```
```
### what follows is the whole content of the file;
### it does not need to be executable;

### to use tmpfs for /home/<user>/.cache/
if [[ ! -e /tmp/kde_cache/ ]]; then
  mkdir /tmp/kde_cache/
  ### the following two commands are because keepassxc currently stores settings under
  ###   /home/<user>/.cache/
  ### and we want to preserve them between reboots;
  mkdir /tmp/kde_cache/keepassxc/
  ln -s /home/<user>/.local/share/keepassxc/keepassxc.ini /tmp/kde_cache/keepassxc/keepassxc.ini
  chown --recursive <user>:<user> /tmp/kde_cache/
  chmod --recursive u=rwX,go= /tmp/kde_cache/
fi
```
_____________________________________________________

```
systemctl reboot
```
